src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,190 @@
+            await DownloadFileShareArtifactAsync(downloadRootPath, destPath, defaultParallelCount, cancellationToken, minimatcherFuncs);
+        }
+        
+        private async Task PublishArtifactUsingRobocopyAsync(AgentTaskPluginExecutionContext executionContext, IHostContext hostContext, string dropLocation, string downloadFolderPath, CancellationToken cancellationToken)
+        {
+            executionContext.Output(StringUtil.Loc("PublishingArtifactUsingRobocopy"));
+            using (var processInvoker = hostContext.CreateService<IProcessInvoker>())
+            {
+                // Save STDOUT from worker, worker will use STDOUT report unhandle exception.
+                processInvoker.OutputDataReceived += delegate (object sender, ProcessDataReceivedEventArgs stdout)
+                {
+                    if (!string.IsNullOrEmpty(stdout.Data))
+                    {
+                        executionContext.Output(stdout.Data);
+                    }
+                };
+
+                // Save STDERR from worker, worker will use STDERR on crash.
+                processInvoker.ErrorDataReceived += delegate (object sender, ProcessDataReceivedEventArgs stderr)
+                {
+                    if (!string.IsNullOrEmpty(stderr.Data))
+                    {
+                        executionContext.Error(stderr.Data);
+                    }
+                };
+
+                var trimChars = new[] { '\\', '/' };
+
+                dropLocation = Path.Combine(dropLocation.TrimEnd(trimChars));
+                downloadFolderPath = downloadFolderPath.TrimEnd(trimChars);
+
+                string robocopyArguments = "\"" + dropLocation + "\" \"" + downloadFolderPath + "\" * /E /COPY:DA /NP /R:3";
+
+                int exitCode = await processInvoker.ExecuteAsync(
+                        workingDirectory: "",
+                        fileName: "robocopy",
+                        arguments: robocopyArguments,
+                        environment: null,
+                        requireExitCodeZero: false,
+                        outputEncoding: null,
+                        killProcessOnCancel: true,
+                        cancellationToken: cancellationToken);
+
+                executionContext.Output(StringUtil.Loc("RobocopyBasedPublishArtifactTaskExitCode", exitCode));
+
+                if (exitCode >= 8)
+                {
+                    throw new Exception(StringUtil.Loc("RobocopyBasedPublishArtifactTaskFailed", exitCode));
+                }
+            }
+        }
+
+        private async Task DownloadFileShareArtifactAsync(string sourcePath, string destPath, int parallelCount, CancellationToken cancellationToken, IEnumerable<Func<string, bool>> minimatchFuncs = null)
+        {
+            var trimChars = new[] { '\\', '/' };
+
+            // If user has specified a relative folder in the drop, change the drop location itself. 
+            sourcePath = Path.Combine(sourcePath.TrimEnd(trimChars));
```

Comment:
>sourcePath.TrimEnd(trimChars) [](start = 38, length = 29)

this will return a string so what is the need for Path.Combine if you're not combining anything?